### ISSUE TITLE
Snapshot retention policy (instances and storage volumes)

### ISSUE THREAD

**Issue by ghost** [lxc/incus#51]
With #9 making significant changes to the structure of the `inc snapshot` command (and there being more of a possibility of breaking backwards compatibility), I'd like to suggest a feature to increase snapshot retention granularity. The following use cases warrant maintaining a history of snapshots, but not at the same duration or granularity as the server where the incus container/VM is actually running:

* Backup Server: you may want to keep a variety of snapshots going back several months (or even years) but you need a variable level of granuarlity as you go farther back in time (or storage space and the sheer number of snapshots will become unmanagable). For example, let's say you take snapshots every hour. You may want a lot of granularity in recent snapshots, say the last 10 hourly snapshots, but then you only need 2 snapshots from last week, 1 snapshot from last month, and 1 from last year. [Sanoid](https://github.com/jimsalterjrs/sanoid) lets you configure this type of variable-granularity-based-on-timescale and it would be great if we could do this with incus too

* Warm Spare: it's desirable to maintain a warm spare server with a copy of the incus container/VM data: in case the active server goes down, the guest can quickly be brought back online on the warm spare. This is discussed briefly [here](https://documentation.ubuntu.com/lxd/en/latest/howto/instances_backup/#copy-an-instance-to-a-backup-server), but utilizing [syncoid](https://github.com/jimsalterjrs/sanoid) and ZFS snapshots to only send the updated delta means that keeping the warm spare up-to-date is easy and doesn't require re-syncing the whole dataset. I think incus supports copying the guest with the `--refresh` flag, but again you have the problem of not being able to maintain a separate snapshot retention policy on the warm spare

Sanoid now supports [pre-snapshot, post-snapshot, and post-prune hooks](https://github.com/jimsalterjrs/sanoid#sanoid-script-hooks) so if incus had a command like `incus snapshot {create,delete} --metadata-only --snapshot-name="xxx"` to only create/delete the database entry but not actually try to `zfs snapshot`/`zfs destroy` the snapshot, this would work (you could let sanoid be in charge of creating and deleting the actual ZFS snapshots). 

Is it possible to add the above type of enhanced functionality for snapshots to incus, or provide some type of integration with 3rd-party tools like sanoid/syncoid?

**Comment by benginow** [comment]
Hi! I'm a UT student working on a group project and would love to take on this issue if it’s still available

**Comment by stgraber** [comment]
Hey @benginow,

This issue is currently marked as a maybe as it's not clear whether what's requested here can actually be made to fit Incus' design.

So I wouldn't recommend starting with this particular issue :)

**Comment by amartin3225** [comment]
I'd be happy to discuss this in more detail if there's anything about this request that could/should be tweaked to make it align better with Incus' design

**Comment by stgraber** [comment]
So in general, I'm definitely not in favor of adding support for creating metadata-only snapshots as without something immediately filling them it effectively leads to a broken database state which may prevent Incus from performing needed data transitions on update or even routine background tasks that span all instances and their snapshots.

If you absolutely want to go down that path, your best bet is to basically do:
```
incus snapshot create INSTANCE SNAPSHOTNAME
zfs destroy POOL/containers/INSTANCE@SNAPSHOTNAME
```

That should result in the same thing but without us having to actually support it ;)


Snapshot retention is also a bit tricky because our fundamental design is based around each snapshot having an expiry date. That's a bit different from backup systems where that's often not true and instead rely on an overall snapshot retention policy to consider the current snapshot set and trim it based on the policy.

It'd be interesting to figure out if given say:
 - snapshot.retention.day=4 (keep max 4 snapshots for the past 24h period)
 - snapshot.retention.week=3 (keep max 3 snapshots for the past 7 days period, excluding current day)
 - snapshot.retention.month=4 (keep max 4 snapshots for the past month, excluding current day)
 - snapshot.retention.year=10 (keep max 10 snapshots for the past year, excluding current month)

It would be possible to already assign an expiry date on the snapshot at the time it's created by basically looking at the available snapshots at the time a new snapshot is created to determine how long it should be kept overall.

If that's possible to do and leads to a useful result, then I think we could implement that.


If not, then the best option would be for 3rd party backup tools that want to manage snapshots and backups on Incus to trigger expiry-less snapshots themselves and then delete them based on their own policy.

**Comment by amartin3225** [comment]
```quote
It'd be interesting to figure out if given say:

snapshot.retention.day=4 (keep max 4 snapshots for the past 24h period)
snapshot.retention.week=3 (keep max 3 snapshots for the past 7 days period, excluding current day)
snapshot.retention.month=4 (keep max 4 snapshots for the past month, excluding current week)
snapshot.retention.year=10 (keep max 10 snapshots for the past year, excluding current month)
It would be possible to already assign an expiry date on the snapshot at the time it's created by basically looking at the available snapshots at the time a new snapshot is created to determine how long it should be kept overall.
```

I think this would work. The only case this doesn't cover is differing retention policies across servers. For example, if you host a container with a busy database, you might not want to keep any snapshots from more than a couple days ago on the running Incus server itself due to the size of the snapshots, but the backup server (which may have more space), could keep additional snapshots. Would these `snapshot.retention.<period>` attributes be set per instance (and be replicated with `incus copy --refresh` to other Incus servers) or globally per Incus server?

```quote
If not, then the best option would be for 3rd party backup tools that want to manage snapshots and backups on Incus to trigger expiry-less snapshots themselves and then delete them based on their own policy.
```

I think the problem with this route, at least historically, is if you delete the ZFS snapshot out from under Incus (e.g. sanoid cleans up a snapshot and then tries to run `incus snapshot delete` as a post-prune hook to clean up the metadata), Incus fails with an error (understandably since it can't find the snapshot it expects to find in the zpool). I suppose an alternative way to handle this would be with some type of `incus snapshot clean` command that would check all snapshots and delete any from the database which don't have a corresponding snapshot in ZFS anymore (e.g. something else modified/deleted it). That would allow a 3rd-party tool to modify the snapshots and then let Incus "catch up" with those changes.

**Comment by stgraber** [comment]
Those proposed config keys  would be instance settings, you could make them apply to multiple instances through  profiles.

For the `--refresh` case with a remote server, differing retention settings based on what I proposed above will not really work since `--refresh` will overwrite the instance config to match the source, though using a profile would avoid that issue. But more importantly because those config keys will just be used to calculate the correct snapshot expiry, a different policy on the target won't have any effect.

One thing that I think we could do is `incus copy --refresh --new-snapshots-only` or something along those lines, so only transferring the newer snapshots and not backfill anything the target may have deleted. Again, that's of limited use since the expiry of the snapshots are pre-calculated, so you'd still need something to trim snapshots after the transfer, but at least you'd only need to trim the new ones and not constantly battle the old ones being copied again.


I'd consider `incus snapshot delete INSTANCE SNAP` failing due to the snapshot already being gone as a bit of a bug and something we could fix (assuming we can accurately differentiate an error about a missing snapshot from another more important error).

**Comment by amartin3225** [comment]
> I'd consider incus snapshot delete INSTANCE SNAP failing due to the snapshot already being gone as a bit of a bug and something we could fix (assuming we can accurately differentiate an error about a missing snapshot from another more important error).

This would be great if possible (and useful even irrespective of this feature request).

The proposed instance config keys along with `incus copy --refresh --new-snapshots-only` implements most of this desired capability (except for the ability to have a different retention on a backup server), so I think it's sufficient to meet this need.

**Comment by stgraber** [comment]
Updated the issue to focus on those proposed additional config keys for both instances and storage volumes:
 - snapshot.retention.day=4 (keep max 4 snapshots for the past 24h period)
 - snapshot.retention.week=3 (keep max 3 snapshots for the past 7 days period, excluding current day)
 - snapshot.retention.month=4 (keep max 4 snapshots for the past month, excluding current week)
 - snapshot.retention.year=10 (keep max 10 snapshots for the past year, excluding current month)

As mentioned above, those will effectively have to conflict with `snapshot.expiry` and will be used to determine an ultimate snapshot expiry date at the time of snapshot creation based on pre-existing snapshots on the instance or storage volume.

**Comment by stgraber** [comment]
I've filed https://github.com/lxc/incus/issues/745 for the other part of this issue.

The last thing mentioned here which are issues with deleting snapshots from Incus when the underlying storage-level snapshot is gone already should be filed separately if that still occurs today.

**Comment by nfournil** [comment]
Nice feature ! I'm doing this via cron script every day ! I had a nice to have feature for this retention policy : after delete account a customer has a legal delay to ask for their data so we must keep these data for several months. In Ceph deleting a volume on a mirror system causes a delete on the destination mirror (and stop mirroring is deleting image also). incus delete image by default. It might be nice to have a switch to move to trash policy on delete. Trash in ceph has a policy to remove from trash after XXX days (so it's perfect !). Difference is : "rbd image rm" ... became "rbd trash mv ...  " .

**Comment by stgraber** [comment]
Yeah, we may be able to have the Ceph driver do that, we'd just need to validate Ceph's behavior when a name gets reused, basically on how to name the trash entries so conflicts can be avoided.

**Comment by kevans1124** [comment]
Hello, I am a UT student working with a partner doing open source contributions for class credit, can my group be assigned this issue please? Or is this still an issue that would not be good to start working with? Thank you!

**Comment by stgraber** [comment]
done!

**Comment by stgraber** [comment]
@kevans1124 are you still working on this?

**Comment by bensmrs** [comment]
I’m not sure how we expect to be able to precompute an expiry time based on these data. Or at least, how it’d be enough. If we manually create a snapshot that goes beyond the limits fixed by the given keys, then, in addition to the precomputed value, we should also remove the oldest snapshot in the given “bucket“ (day/week/month/year).

Another issue arises: what if I just want day and a year policies? Then the logic excluding the current week/month would be inexact, as there’s no policy covering them. Forcing the operator to set all 4 keys doesn’t seem reasonable, as there are genuine use cases where we don’t want such a fine granularity.

Just adding a coin to the machine, would it be interesting to have, in addition to these simple keys, a scriptlet-based approach that takes `(name, timestamp)` pairs and returns the snapshots to keep/remove? Maybe along with a key indicating how often the scriptlet should be run.

**Comment by stgraber** [comment]
Just spent quite a bit of time thinking through this one again and there's clearly no way to make it work as described without having to rewrite the expiry of existing snapshots, which kinda defeats the purpose of this lightweight approach and gets us into actual snapshot scheduling territory.

For more complex snapshot scheduling, we've typically recommended driving both the snapshot and backup APIs from an external script over the API as that's what will provide the most flexibility and will allow actually enforcing policies across all projects rather than leave it up to the individual operator to get right.

Because of security concerns with scriptlets, they need to be restricted to users who can already get system wide admin access as they can be used to DoS a system and so are typically unsuitable for features that are exposed to all users.

So a global scriptlet handling automatic snapshotting across all instances on the system could work, but something at the instance or project level, not so much.


The general ideas as far as retention policy from this issue still generally make sense I think, but rather than be able to compute an actual expiry, we'd need to have a flag on the snapshot telling us that it was automatically created and is managed by the retention policy.

We'd still use the normal scheduling process but without expiry and effectively check how many snapshots we have for each retention bucket, then when needed delete the oldest entry in a given bucket to make space for a new one.

**Comment by bensmrs** [comment]
> Just spent quite a bit of time thinking through this one again and there's clearly no way to make it work as described without having to rewrite the expiry of existing snapshots, which kinda defeats the purpose of this lightweight approach and gets us into actual snapshot scheduling territory.

Well unfortunately I think it’s one of these cases where to implement simple APIs for the user, a whole new codebase has to emerge…

> Because of security concerns with scriptlets, they need to be restricted to users who can already get system wide admin access as they can be used to DoS a system and so are typically unsuitable for features that are exposed to all users.

TBF, in this case, spawning a thread with a 5s timeout (or even shorter, as I don’t expect decisions to be CPU-intensive) doesn’t feel too dangerous.

> The general ideas as far as retention policy from this issue still generally make sense I think, but rather than be able to compute an actual expiry, we'd need to have a flag on the snapshot telling us that it was automatically created and is managed by the retention policy.

I’m feeling quite a few untold corner cases there :)
And I’m not sure this flag really is needed; are you considering that manually-triggered snapshots shouldn’t count in the retention policy?

**Comment by stgraber** [comment]
> I’m feeling quite a few untold corner cases there :)\nAnd I’m not sure this flag really is needed; are you considering that manually-triggered snapshots shouldn’t count in the retention policy?

I think they should count but they absolutely should not ever be automatically deleted, so that's why we need to know what created the snapshot.

**Comment by stgraber** [comment]
> TBF, in this case, spawning a thread with a 5s timeout (or even shorter, as I don’t expect decisions to be CPU-intensive) doesn’t feel too dangerous.

I don't know how we would apply such restrictions on the sandbox or the goroutine running it. We can't really do it at the OS level as there is no guaranteed mapping between threads and goroutines.

The only safe way I can think of would be to fork out a subprocess but given systems with thousands of instances, that could become quite expensive.

**Comment by bensmrs** [comment]
As things are, the current description is still a bit too underspecified. There are basically 4 approaches that I just studied:

* Buckets are not cumulative (as proposed), and defining a bucket limit imposes to define smaller bucket limits (i.e. defining a `month` limit requires to define `week` and `day` limits)
  * Pros: easy to define, stable semantics, easy to predict
  * Cons: big weight on sysadmins, requires to define granularities that sometimes don’t matter (i.e. impossible to only define `day` and `month`).
* Buckets are not cumulative (as proposed), and smaller undefined buckets have unlimited sizes (i.e. defining a `week` limit without a `day` limit allows to create as many `day` snapshots, which will effectively be filtered by the `week` limit the day after)
  * Pros: easy to define, stable semantics
  * Cons: hard to predict (as effects can hit hard if someone went yolo on `day` buckets, and it ends up deleting all the week snapshots 24h later), can grow in size unexpectedly
* Buckets are not cumulative (as proposed), and smaller undefined buckets are included in the smallest bigger bucket (i.e. defining a `week` limit without a `day` limit includes the `day` bucket in the `week` bucket)
  * Pros: easy to define, easy to predict
  * Cons: unstable semantics, can create damage in case of mistakes (i.e. having a `day` limit of 10 and a `week` limit of 10, if one deletes the `day` limit, then suddenly the 10 `day` snapshots are included in the `week` bucket, thus deleting all the older `week` snapshots)
* Buckets are cumulative (i.e. retentions are defined cumulatively and bigger buckets include smaller ones)
  * Pros: easy to define, stable semantics, easy to predict, removing a limit by accident causes absolutely no damage (as the bigger key is left unchanged)
  * Cons: ~~sysadmins need to know how additions work~~, manually-defined snapshots count for all buckets (which can be problematic if someone goes yolo creating 100 snapshots manually; possible solution at the end of the proposal).

To me, buckets must be cumulative, allowing non-destructive mistakes, along with a uniform behavior and a clear specification.

So, to summarize (with some reinterpretation and changes to make development simpler),
* The user can define any combination of `snapshot.retention.{day,week,month,year}`;
* An invariant guarantees that `snapshot.retention.day` ⩽ `snapshot.retention.week` ⩽ `snapshot.retention.month` ⩽ `snapshot.retention.year`;
* $B_{\mathtt{day}}\subseteq B_{\mathtt{week}}\subseteq B_{\mathtt{month}}\subseteq B_{\mathtt{year}}$
* For $i\in\left\\{\mathtt{day},\mathtt{week},\mathtt{month},\mathtt{year}\right\\}$ do
  * If $n := \\#B_i - \mathtt{snapshot.retention.}i > 0$ then
    * Remove the oldest $n$ snapshots from $\overline{B_i}:=\left\\{s\mid s\in B_i\land s\mathrm{\ automatically\ created}\right\\}$
    * If $n>\\#\overline{B_i}$ then remove all $s\in\overline{B_i}$ and warn that some manually created snapshots were not deleted with the given policy, despite being above the threshold

A few questions:
* Where would the automatic/manual flag be stored? a new DB column in `storage_volumes_snapshots`?
* When/Where is the bucket checking logic triggered?
* Can we use the smallest bucket limit to actually limit manual snapshot creation? I’m thinking of: manually creating a snapshot that goes above the lowest limit would trigger the deletion of the oldest automatic snapshot in the smallest bucket, and, if there are none, allow the deletion of the oldest manual snapshot in this bucket. Because the operation is manual, I think we can do it this way. It may require an additional API parameter like `force` or `delete_old_manual_snapshots` to allow old manual snapshot removal on new manual snapshot creation, else, we can fail the operation with an error (suggesting to use the additional API parameter/CLI flag).

If you agree with the global idea, I’m interested in handling the implementation.

**Comment by stgraber** [comment]
> Where would the automatic/manual flag be stored? a new DB column in `storage_volumes_snapshots`?

Yeah, see #2254, we need to do some DB cleanup of snapshot handling anyways so we can decouple the snapshot itself from the data it contains. Once we have that, we can easily add this as a `trigger` or similar column in the table.

> When/Where is the bucket checking logic triggered?

I'd do it through the normal scheduling logic, so when a snapshot is scheduled to be taken automatically, it gets made and then the bucket logic is ran to re-shuffle snapshots as needed.

> Can we use the smallest bucket limit to actually limit manual snapshot creation? I’m thinking of: manually creating a snapshot that goes above the lowest limit would trigger the deletion of the oldest automatic snapshot in the smallest bucket, and, if there are none, allow the deletion of the oldest manual snapshot in this bucket. Because the operation is manual, I think we can do it this way. It may require an additional API parameter like `force` or `delete_old_manual_snapshots` to allow old manual snapshot removal on new manual snapshot creation, else, we can fail the operation with an error (suggesting to use the additional API parameter/CLI flag).

If we have the distinction between manually triggered and automatically triggered snapshots, I'd say we just leave the manually triggered ones alone, those should not count towards any of the retention buckets and will appear clearly as having been manually created and therefore up to the user to delete either manually or through an expiry.

Am I missing something?

**Comment by bensmrs** [comment]
> If we have the distinction between manually triggered and automatically triggered snapshots, I'd say we just leave the manually triggered ones alone, those should not count towards any of the retention buckets and will appear clearly as having been manually created and therefore up to the user to delete either manually or through an expiry.
>
>Am I missing something?

I actually still had this previous conversation in mind:

>> are you considering that manually-triggered snapshots shouldn’t count in the retention policy?
>
> I think they should count but they absolutely should not ever be automatically deleted, so that's why we need to know what created the snapshot.

Are we in line with the rest of the proposal?

**Comment by stgraber** [comment]
Ah yeah, I guess I changed my mind on that one after reading through your pros/cons of the various approaches :)


Sounds like the way to go is with the cumulative approach but only apply the logic to snapshots created through scheduling, completely leaving alone anything that the user created directly themselves.

**Comment by bensmrs** [comment]
> Ah yeah, I guess I changed my mind on that one after reading through your pros/cons of the various approaches :)

Ok then, let’s go for it and stick with it. I’ll consider #2254 to be a blocker for now.

**Comment by heldergomes** [comment]
I'm not sure if someone else is working on this feature, but I got really interested to working on it!
If you guys are not in rush to delivery it, I could take it!

**Comment by stgraber** [comment]
@bensmrs showed some interest to tackle this one but I don't know how things look like timeline wise on his end. Though as mentioned, we'd probably want to sort out #2254 first so we have the database in a state that's compatible with implementing this feature.

**Comment by bensmrs** [comment]
Yeah I’m still interested in working on this one, but #2254 really is a blocker (and that one is gonna take quite some time, so I’m not sure to work on it).

**Comment by heldergomes** [comment]
It's ok, I could work on #2254 instead, if @bensmrs is not working on that issue!

If so, I'd like to ask for a brief help about which components I should digging deep to understand how to make those changes

**Comment by bensmrs** [comment]
Oh sure, please have fun with that one (and feel free to ask for help on the issue discussion)
